File: Go/HDBn/HDBn.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
)

const n int = 3

func readFile(f *os.File) []byte {
	line := bufio.NewReader(f)
	source, err := line.ReadBytes('\n')
	if err != nil && err != io.EOF {
		panic(err)
	}
	return source
}

func encode(source []byte) []byte {
	var result = make([]byte, 0)
	counter := 0
	last := 1
	lastV := 1
	for i := 0; i < len(source); i++ {
		if source[i] == '1' {
			counter = 0
			if last == 1 {
				result = append(result, 'n')
			} else {
				result = append(result, '1')
			}
			last = -last
		} else if source[i] == '0' {
			counter++
			result = append(result, '0')
			if counter == n+1 {
				counter = 0
				if lastV == 1 {
					result[i] = 'n'
				} else {
					result[i] = '1'
				}
				lastV = -lastV
				if result[i-n-1] != result[i] {
					result[i-n] = result[i]
				}
			}
		}
	}
	return result
}

func decode(coding []byte) []byte {
	result := make([]byte, 0)
	last := 1
	for i := 0; i < len(coding); i++ {
		if coding[i] == 'n' {
			result = append(result, '1')
			if last == -1 {
				for j := i; j >= i-n; j-- {
					result[j] = '0'
				}
			} else {
				last = -last
			}
		} else if coding[i] == '1' {
			result = append(result, '1')
			if last == 1 {
				for j := i; j >= i-n; j-- {
					result[j] = '0'
				}
			} else {
				last = -last
			}
		} else if coding[i] == '0' {
			result = append(result, '0')
		}
	}
	return result
}

func main() {
	f, err := os.Open("HDBn/test.txt")
	if err != nil {
		fmt.Printf("file is not exist")
		os.Exit(1)
	}
	source := readFile(f)
	coding := encode(source)
	decoding := decode(coding)
	fmt.Println("source:  ", string(source))
	fmt.Println("coding:  ", string(coding))
	fmt.Println("decoding:", string(decoding))
}

File: Python/HDBn/HDBn.py
# n代表-1

import os

N = 3

def decode(coding):
    result = []
    last = 1
    for i in range(len(coding)):
        if coding[i] == 'n':
            result.append('1')
            if last == -1:
                for j in range(i, i - N - 1, -1):
                    result[j] = '0'
            else:
                last = -last
        elif coding[i] == '1':
            result.append('1')
            if last == 1:
                for j in range(i, i - N -1, -1):
                    result[j] = '0'
            else:
                last = -last
        elif coding[i] == '0':
            result.append('0')
    return ''.join(result)

def encode(source):
    result = []
    # 第一个“1”被编码为“-1”
    # 第一个“V”被编码为“-V”
    counter, last, lastV = 0, 1, 1
    for i in range(len(source)):
        if source[i] == '1':
            counter = 0
            result.append('n' if last == 1 else '1')
            last = -last
        elif source[i] == '0':
            counter += 1
            result.append('0')
            if counter == N + 1:
                counter = 0
                result[i] = 'n' if lastV == 1 else '1'
                lastV = -lastV
                if result[i - N - 1] != result[i]:
                    result[i - N] = result[i]
    return ''.join(result)

def main():
    if os.path.exists("Python/HDBn/test.txt"):
        with open(r"Python\HDBn\test.txt", "r") as f:
            source = f.readline()
            coding = encode(source)
            decoding = decode(coding)
            print(f"source:\t\t{source}")
            print(f"coding:\t\t{coding}")
            print(f"decoding:\t{decoding}")
    else:
        print("file is not exist")

if __name__ == "__main__":
    main()